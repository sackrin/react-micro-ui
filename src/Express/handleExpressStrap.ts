import path from 'path';
import express from 'express';
import { createElement } from 'react';
import embedComponent from '../Helpers/embedComponent';
import type { MicroUiConfig } from "../Types/MicroUiConfig";
import type { MicroUiConfigProfileEnv } from "../Types/MicroUiConfigProfileEnv";

type HandleExpressStrap = (
  name: string,
  component: string,
  env: MicroUiConfigProfileEnv,
  config: MicroUiConfig,
  method: string,
) => (req: express.Request, res: express.Response) => Promise<void>;

// Direct Import React
// We have to do it this way to permit SSR react + hooks
const ReactDOMServer = require(path.join(process.cwd(), 'node_modules', 'react-dom', 'server'));

const handleExpressStrap: HandleExpressStrap = (name, component, env, config, method) => async (req, res) => {
  const props = method === 'GET' ? req.query : req.body;
  const rendered = ReactDOMServer.renderToString(createElement(component, props));
  const html = embedComponent(name, env, config, props, rendered);
  res.set('Content-Type', 'text/html');
  res.send(html);
  res.sendStatus(200);
};

export default handleExpressStrap;
