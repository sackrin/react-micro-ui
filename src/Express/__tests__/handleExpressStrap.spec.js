import React from 'react';
import { expect } from 'chai';
import sinon from 'sinon';
import handleExpressStrap from '../handleExpressStrap';
import config from '../../microui.default.config';

describe('Express/handleExpressStrap', () => {
  const ExampleComponent = props => <div>{props.name}</div>;

  it('can correctly out a server side rendered component with GET variables', async () => {
    const mockReq = {
      query: { name: 'John Cater' },
      body: {},
    };
    const mockRes = {
      set: sinon.fake(),
      send: sinon.fake(),
      sendStatus: sinon.fake(),
    };
    await handleExpressStrap('ExampleComponent', ExampleComponent, console, {}, config, 'GET')(mockReq, mockRes);
    expect(mockRes.set).to.have.been.calledWith('Content-Type', 'text/html');
    jestExpect(mockRes.send).toMatchSnapshot();
    expect(mockRes.sendStatus).to.have.been.calledWith(200);
  });

  it('can correctly out a server side rendered component with POST variables', async () => {
    const mockReq = {
      query: {},
      body: { name: 'Mark Green' },
    };
    const mockRes = {
      set: sinon.fake(),
      send: sinon.fake(),
      sendStatus: sinon.fake(),
    };
    await handleExpressStrap('ExampleComponent', ExampleComponent, console, {}, config, 'POST')(mockReq, mockRes);
    expect(mockRes.set).to.have.been.calledWith('Content-Type', 'text/html');
    jestExpect(mockRes.send).toMatchSnapshot();
    expect(mockRes.sendStatus).to.have.been.calledWith(200);
  });
});
