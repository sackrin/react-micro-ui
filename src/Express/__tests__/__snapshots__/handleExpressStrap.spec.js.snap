// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Express/handleExpressStrap can correctly out a server side rendered component with GET variables 1`] = `[Function]`;

exports[`Express/handleExpressStrap can correctly out a server side rendered component with POST variables 1`] = `[Function]`;
